### ISSUE TITLE
PeerTube as Hidden Service

### ISSUE THREAD

**Issue by 6543** [Chocobozzz/PeerTube#1569]
I like to setup PeerTube as Hidden service wich can federate to other Tor-PeerTube instances.
For now i have two wich work perfect as Hidden service - but as standalone.

There are Two Issues:
- How to set a proxy for outgoing trafic?
- how to disable the requirement of ssl to follow each other (tor trafic is already E2E-Encrypted)

**Comment by 6543** [comment]
an idear/notize:

what about optional tor integration for normal nodes?
add an optional "tor gateway" to the config and make the case if a node has "onion" as TLD then it shoud do all routines as usual but use tor.

**Comment by buoyantair** [comment]
The first concern that comes to my mind as soon as I hear about peertube on tor is,
if you are doing it for privacy concerns, torporject openly expresses concern over using bittorrent on tor.

please read more about it here -> https://blog.torproject.org/bittorrent-over-tor-isnt-good-idea

**Comment by 6543** [comment]
This is because lot of torrent clients leak ip addresses even it a proxy is set.
And ZeroNet uses Tor successfuly with torren too. So it only depends on how you implement it.

**Comment by buoyantair** [comment]
@6543 I agree. But in such a case, will I be able to watch the video hosted on the tor network from an instance hosted on the normal internet?

**Comment by rigelk** [comment]
@buoyantair probably not. As it stands, to watch videos served over Tor your browser should be able to resolve onion addresses, or use an instance able to do so as a proxy.

First option is going to cause errors on most clients. Either we consider that too much to handle or we deal with the error gracefully/separate these videos from the rest.

While possible, the second option is already using PeerTube for some usage outside of the classic HTTP server territory, so it falls in the "yes but as a plug-in" category.

**Comment by 6543** [comment]
but for now, how do i just use a sock proxy? To test the node behaviour

**Comment by 6543** [comment]
@rigelk there wont be "bronek video entrys".
I have tested to "follow" a hidden service and otherwise, but they dont conect. On the hidden one the entry stay "Pending"...

So for now it would be nice to just conect  hidden once and then we can think to conect both worlds?

**Comment by buoyantair** [comment]
@rigelk But then again, lets an instance is federating over tor, and another is federating over https, and they both connect to each other, in which case I can connect to the instance on https and there by access the video from tor network right? (as a peer?)

**Comment by rigelk** [comment]
@buoyantair right now you wouldn't be able to do that, no. PeerTube is not meant to act as a proxy.

**Comment by buoyantair** [comment]
@rigelk Wouldnt that fracture the network then?

**Comment by Houkime** [comment]
While Peertube itself is not meant to run as a proxy, one can still write a specialised server which _will_ act as a proxy and represent the Tor part of federation for the clearnet part?
But in this case, how could many such bridges coexist together without appearance of duplicate videos?
Does one need to make changes to protocols for such bridges to work effectively and balance the load?

**Comment by 6543** [comment]
If we just add config enable tor wich three options: [disable/enable/only]

If disabled PeerTube will ignore all federation requests from .onion
if enabled Peertube uses SOCK Proxy for .onion hostnames
If only it will forward all requests throu SOCK Proxy

how hard is it to use diffetent networks with typescript? and differ federation routines based on TLDs?

**Comment by trymeouteh** [comment]
Would be great for a peertube server to be able to talk to onion sites and allow anyone with a peertube account on the clear web to watch and connect to a video on an onion site.

**Comment by Chocobozzz** [comment]
Merging this issue in https://github.com/Chocobozzz/PeerTube/issues/1886
